### PR TITLE
Implement query support for null

### PR DESCRIPTION
**Pull request by tgoyne**
@bdash 

Main WIP part is that there's a lot more stuff to test, but I didn't want to wait too long on throwing this up in case it conflicts with the other query work.

### CODE REVIEW

File: Realm/RLMQueryUtil.mm [hunk 1]
```diff
@@ -538,6 +517,43 @@ void add_constraint_to_query(realm::Query &query, RLMPropertyType type,
     }
 }
 
+void do_add_constraint_to_query(realm::Query&, RLMPropertyType, NSPredicateOperatorType,
+                                NSComparisonPredicateOptions, const std::vector<NSUInteger>&,
+                                id, realm::null)
+{
+    // This probably isn't actually reachable since there's earlier checks for
+    // it, but they're done before null is un-type-erased
+    @throw RLMPredicateException(@"Invalid predicate expressions",
```
Comment (bdash) [inline Realm/RLMQueryUtil.mm]: This definitely isn't reachable due to the logic in `update_query_with_predicate`. I'm guessing this overload is necessary in order for `add_constraint_to_query` to compile?

File: Realm/RLMQueryUtil.mm [hunk 2]
```diff
@@ -401,73 +405,48 @@ void process_or_group(Query &query, id array, Func&& func) {
...
-template <typename TableGetter>
-struct ValueOfTypeHelper<String, TableGetter> {
-    static id convert(TableGetter&&, id value)
-    {
-        return value;
-    }
+template <typename RequestedType, typename TableGetter>
+struct ValueOfTypeHelper {
+    static realm::null convert(TableGetter&&, realm::null) { return realm::null(); }
+    static auto convert(TableGetter&&, id value) { return ::convert<RequestedType>(value); }
```
Comment (bdash) [inline Realm/RLMQueryUtil.mm]: This is much tidier!

**Comment by bdash** [comment]
It'll conflict with the key-path vs key-path changes I'm working on, but I can work that out when we pick up a new enough version of core to support them.

:+1:

**Comment by tgoyne** [comment]
Added basic tests for every supported operator on every nullable property type.

File: Realm/Tests/QueryTests.m [hunk 3]
```diff
@@ -1929,4 +1922,269 @@ - (void)testSortingColumnsWithNull {
...
+        {@"doubleObj", @"1", @"0", @1, @0, true},
+        {@"string", @"'a'", @"''", @"a", @"", false, true},
+        {@"data", nil, nil, notMatchingData, matchingData},
+        {@"date", nil, nil, notMatchingDate, matchingDate, true},
+    };
+
+    // Assert that the query "prop op value" gives expectedCount results when
+    // assembled via string formatting
+#define RLMAssertCountWithString(expectedCount, op, prop, value) \
+    do { \
```
Comment (jpsim) [inline Realm/Tests/QueryTests.m]: why are these defines all wrapped in a `do {} while(0)`?

Comment (tgoyne) [inline Realm/Tests/QueryTests.m]: It makes the macro expand to the equivalent of a single statement that requires a trailing semicolon, which avoids a bunch of potential bugs from macro-expansion edge cases.

Comment (jpsim) [inline Realm/Tests/QueryTests.m]: fair enough, thanks!

**Comment by bdash** [comment]
:+1:

File: Realm/RLMQueryUtil.mm [hunk 4]
```diff
@@ -538,6 +519,43 @@ void add_constraint_to_query(realm::Query &query, RLMPropertyType type,
...
+void add_constraint_to_query(realm::Query &query, RLMPropertyType type,
+                             NSPredicateOperatorType operatorType,
+                             NSComparisonPredicateOptions predicateOptions,
+                             const std::vector<NSUInteger>& linkColumns, L lhs, R rhs)
+{
+    // The expression operators are only overloaded for realm::null on the rhs
+    RLMPrecondition(!is_nsnull(lhs), @"Unsupported operator",
+                    @"Nil is only supported on the right side of operators");
+
+    if (is_nsnull(rhs)) {
```
Comment (jpsim) [inline Realm/RLMQueryUtil.mm]: this could be simplified a bit by using `auto validatedRHS = is_nsnull(rhs) ? realm::null() : rhs;` and passing the validated value to `do_add_constraint_to_query`.

Comment (tgoyne) [inline Realm/RLMQueryUtil.mm]: That won't compile because `rhs` and `realm::null()` are unrelated types that can't be converted to each other. The whole point of this is to call different overloads (ultimately of the core query functions) depending on the runtime value of `rhs`.

Comment (jpsim) [inline Realm/RLMQueryUtil.mm]: ah, yes makes sense
